### ISSUE TITLE
Pass options to models and collections?

### ISSUE THREAD

**Issue by bazineta** [dhruvaray/backbone-associations#84]
I must be missing something obvious here.

Is there a way to specify options to be passed to the models and collections created, i.e., that which ends up being the second parameter to a model or collection's initialize() function?

I have a top-level object which contains an object graph; as one traverses down the graph, the server will respond on URLs corresponding to the nested objects, in a manner similar to that of the incremental save/fetch example.  However, the top-level object is based on a Mongo ID, and thus for the sub-objects to function, I need to be able to provide them with their url, e.g., where id is the object ID provided by the server:

top: /top/id
  nested: /top/id/nested
    collection: /top/id/nested/collection

I looked at the incremental save/fetch example, and it appears that I can't use that, as, in the intialize() function of the top object, the nested object is still just data, not the AssociatedModel, and thus setting the url there as done in the example doesn't work.

I was hoping to be able to do something along the lines of specifying an object or function in the relation hash that would be automatically passed to the initialize function of the collection or model created, so that as the object graph is created, the objects could just pass down the url that the nested objects or collections should use, e.g., in top:

```
relations: [{
  type: Backbone.One,
  key: 'nested',
  relatedModel: Nested,
  modelOptions: {
    url: this.url() + "/nested"
 }
}]
```

and then in Nested:

```
initialize: function (options) {
  this.url = options.url;
}
```

As I started with, I must be missing something fundamental here; any pointers would be appreciated.  Thanks.

**Comment by dhruvaray** [comment]
> > in the intialize() function of the top object, the nested object is still just data, not the AssociatedModel, and thus setting the url there as done in the example doesn't work.

Why do you say this? By the time you come to the initialize method, you should have the `AssociatedModel` instances already created. [`set` (which is responsible for creating the `AssociatedModel` instances) is called before `initialize` in the model constructor in BB] 

From the `Department` `initialize` method 

``` javascript
initialize: function(){
        var dept = this;
        //Not nested data. There is a sub-graph rooted at `Locations`
        this.get('locations').url = function(){
            return '/department/' + dept.id + '/locations';
        }
    },

```

Am I missing something?

**Comment by bazineta** [comment]
Thanks, looking at the recipe again, I see where I was falling into the issue; the collection didn't have a default url set, as it does in the example.  Setting one there, and a urlRoot on the contained objects, allows the subojects to be built properly, and they exist in the initialize routines.

However, the problem is, in my case that's too late to set the urls; the collection objects are already in the collection, and now all have the wrong url.  I think my case may differ from the recipe in that the objects in the graph are purely hierarchical; that is, they must exist under the top level object, whereas in the example, a location with the default url is fine, since in the code in the response above, the result is a collection with a url that doesn't match that of the contained models; i.e.,after the code snippet in the previous response, the locations collection will have a url of /department/id/locations, but the contained objects will have a url of /location/id, based on their urlRoot.

In short the difference is that in the example, a location is a first-class object, so a default url of /location/id is ok.  In my case, a location does not have an independent existence, and thus having a default url isn't usable; rather, the url it ends up with would be relative to the department, e.g., /department/id/locations/id is what I need for the server side.

When I was doing this by hand, I'd get the graph and parse it top-down, so I'd have the proper url to hand to the collection in the collection options, i.e., relative to the department, in the initialization options, so that the objects created would end up with urls based on the collection url, rather than the urlRoot.

**Comment by jdkanani** [comment]
The following code snippet will do what you are looking for:

``` js
// Backup original url function
var originalUrl = Backbone.AssociatedModel.prototype.url;
// override our version of url
_.extend(Backbone.AssociatedModel.prototype, {
    url: function(){
        var base = originalUrl.call(this), // get original url for current model
            parents = this.collection && this.collection.parents;
        // If model's collection has parent, prepend it to base
        if(parents && parents.length === 1){
            base = _.result(parents[0], 'url') + base;
        }
        return base;
    }
});
```

Load this snippet after `backbone-associations.js` and before your model initialization.

Here is the demo code that you can try out:

``` js

var Location = Backbone.AssociatedModel.extend({
    defaults:{
        id:-1,
        add1:"",
        add2:null,
        zip:"",
        state:""
    }
});

var Locations = Backbone.Collection.extend({
    model: Location,
    url: '/locations'
});

var Department = Backbone.AssociatedModel.extend({
    relations:[{
        type:Backbone.Many,
        key:'locations',
        collectionType: Locations
    }],
    defaults:{
        name:'',
        number:-1
    }
});

var Departments = Backbone.Collection.extend({
    model: Department,
    url: '/departments'
});


var departments = new Departments(
    [
        {
            'name': 'dep1',
            'id': 1,
            'locations': [
                {
                    id: 'loc1-1',
                    zip: '91001'
                },
                {
                    id: 'loc1-2',
                    zip: '91002'
                },
                {
                    id: 'loc1-3',
                    zip: '91003'
                }
            ]
        },
        {
            'name': 'dep2',
            'id': 2,
            'locations': [
                {
                    id: 'loc2-1',
                    zip: '91004'
                },
                {
                    id: 'loc2-2',
                    zip: '91005'
                },
                {
                    id: 'loc2-3',
                    zip: '91006'
                }
            ]
        }
    ]
);

console.log(departments.at(0).get('locations[0]').url()); // /departments/1/locations/loc1-1
console.log(departments.at(0).get('locations[2]').url()); // /departments/1/locations/loc1-3

console.log(departments.at(1).get('locations[1]').url()); // /departments/2/locations/loc2-2
console.log(departments.at(1).get('locations[2]').url()); // /departments/2/locations/loc2-3
```

Let me know if it works for you.

**Comment by dhruvaray** [comment]
@bazineta - Did this work for you?

**Comment by bazineta** [comment]
Apologies for not responding to this sooner.  Not as yet in terms of patching the prototype object, but using the same logic in the extended models and collections down the hierarchy works, and is probably more suited for my particular application.  In short, binding it late is the key, and I was unaware of the awesome _.result(), function, thanks for the pointer.

**Comment by dhruvaray** [comment]
Good to hear that... The snippet is a working example. So I am surprised that it did not work for you. In any case, we will pull it as a recipe on our website with a working jsfiddle shortly.

Do star us if you find the library useful in your work!

**Comment by jdkanani** [comment]
Great! I have already created [jsfiddle](http://jsfiddle.net/jdkanani/T2Ggf/) and [gist](https://gist.github.com/jdkanani/a22063f1a9567e21aee4) for this example.

> but using the same logic in the extended models and collections down the hierarchy works, and is probably more suited for my particular application. 

Yup. It will also work.

>   In short, binding it late is the key

Exactly, you also need `parents` property to construct model's `url`.

> I was unaware of the awesome _.result(), function, thanks for the pointer

Enjoy!

**Comment by bazineta** [comment]
Adding to the solution here, a very nice functional mixin plugin:

https://github.com/rhysbrettbowen/Backbone.Advice

Using this approach, one can, for example, handle cases where models contain associated collections, which contain associated models, etc., down an arbitrary hierarchy, while keeping the logic in one spot:

```
 // Define a mixin
 var AssociatedURL = function (options) {
      this.clobber({
          url: function () {
              var base = '/' + options.url;
              if (this.parents && this.parents.length === 1) {
                  base = _.result(this.parents[0], 'url') + base;
              }
              return base;
          }
      })
  };

var Model = Backbone.AssociatedModel.extend({
    defaults: {
        foo: "bar"
    }
}).mixin([ AssociatedURL ], { url: 'model' });

var Collection = Backbone.Collection.extend({
    model: Model,
}).mixin([ AssociatedURL ], { url: 'collection' });
```
